### PR TITLE
Improve performance of Spinlocks

### PR DESCRIPTION
**Pull request by NK-Nikunj**
The current implementation of spinlocks uses atomic exchanges on every loop iteration that can cause excess contention due to cache coherence policies within the CPU. This will lead to overwhelming traffic on the bus reducing the performance of the locks. The backoff through yield does reduce contention but we can still do better in performance.

I have mitigated this by first checking if the lock can be acquired by loading the value to cache and then spinning on it every iteration. This reduces the contention from every iteration to iterations where the lock has been released by another thread. The improvements will be insignificant for low contention situations (probably an improvement of a couple of nanoseconds) but should be observable for very high contention situations.

### CODE REVIEW

File: libs/core/synchronization/include/hpx/synchronization/spinlock.hpp [hunk 1]
```diff
@@ -49,7 +49,30 @@ namespace hpx { namespace lcos { namespace local {
...
+            //      Thread 2 yields while is_locked() computes to
+            //      false. Then it retries doing is_locked() -> false
+            //      followed by an acquire_lock() operation.
+            //      The above order can be changed arbitrarily but
+            //      the nature of execution will still remain the
+            //      same.
+            while (is_locked() || !acquire_lock())
             {
                 util::yield_while([this] { return is_locked(); },
                     "hpx::lcos::local::spinlock::lock");
```
Comment (msimberg) [inline libs/core/synchronization/include/hpx/synchronization/spinlock.hpp:78]: Would this be equivalent, and possibly simpler?

```suggestion
            do 
            {
                util::yield_while([this] { return is_locked(); },
                    "hpx::lcos::local::spinlock::lock");
            } while (!acquire_lock())
```
(ignore the mismatched lines, github doesn't let me comment on lines below the diff).

Can we construct a small test to verify the effect of this? If nothing else, running the `future_overhead` benchmark, which relies on (mostly uncontended) spinlocks.

Thanks for looking into this!

Comment (NK-Nikunj) [reply libs/core/synchronization/include/hpx/synchronization/spinlock.hpp:78]: Good catch! We can certainly do it this way to make it look simpler. This will also ensure that there is a single load operation when leaving the loop (instead of the current 2). I'll correct it right away.

Comment (NK-Nikunj) [reply libs/core/synchronization/include/hpx/synchronization/spinlock.hpp]: @msimberg Let me write a small high contention test to test the performance. I do not suspect any significant speedup but decreasing the CPU bus traffic may potentially help us with performance somewhere else.

Comment (hkaiser) [reply libs/core/synchronization/include/hpx/synchronization/spinlock.hpp]: This change will take effect in the uncontended case mostly. With higher contention false sharing is still an issue. But this can't be resolved that easily. Thanks for looking into this.

Would you mind fixing the `util::spinlock` in a similar way? This needs to be changed to try a load first as well: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/thread_support/include/hpx/thread_support/spinlock.hpp#L48.

Comment (NK-Nikunj) [reply libs/core/synchronization/include/hpx/synchronization/spinlock.hpp]: @hkaiser the changes in the commit also include changes to the link you directed to, or am I getting something wrong?

Comment (hkaiser) [reply libs/core/synchronization/include/hpx/synchronization/spinlock.hpp]: Perfect, I didn't see that. Thanks!

File: libs/core/thread_support/include/hpx/thread_support/spinlock.hpp [hunk 2]
```diff
@@ -45,15 +45,18 @@ namespace hpx { namespace util { namespace detail {
...
             // Wait for lock to be released without generating cache misses
+            // Similar implementation to hpx::lcos::local::spinlock
             unsigned k = 0;
-            do
+            while (m.load(std::memory_order_relaxed) ||
+                m.exchange(true, std::memory_order_acquire))
             {
                 yield_k(k++);
-            } while (!try_lock());
+            }
```
Comment (hkaiser) [inline libs/core/thread_support/include/hpx/thread_support/spinlock.hpp:59]: A `do{} while()` would be more appropriate as we know that the lock can't be acquired if `k == 0`. Also, wy not call `try_lock()` instead?

Comment (NK-Nikunj) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp:59]: `while()` vs `do{} while()` : `yield_k` for `k == 0` is a no-op so I felt both ways of representing are equivalent. This can be changed however if you'd like it to be a `do{} while()` without any change in performance.

why not call `try_lock`? The nature of the return statement means that `!m.load(std::memory_order_relaxed)` will always compute to false in cases where the lock has been previously acquired. This means that `!try_lock()` will essentially trigger an exchange statement every iteration. This can be mitigated by doing something like:

```
while (true)
{
    if (try_lock())
        return;
    yield_k(k++);
}
```

Would you want me to change the code to look like above?

Comment (NK-Nikunj) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp:59]: It could also be turned into a `do{} while()` using which we can get rid of the above optimistic try_lock. I think I'll convert it to `do{} while()` for better code clarity.

File: libs/core/thread_support/include/hpx/thread_support/spinlock.hpp [hunk 3]
```diff
@@ -44,19 +45,17 @@ namespace hpx { namespace util { namespace detail {
...
             {
+                // Wait for lock to be released without generating cache misses
+                // Similar implementation to hpx::lcos::local::spinlock
+                if (try_lock())
+                    return;
+
                 yield_k(k++);
-            }
+
+            } while (true);
```
Comment (hkaiser) [inline libs/core/thread_support/include/hpx/thread_support/spinlock.hpp]: What's wrong with simply writing:
```
do {
    yield_k(k++);
} while (!try_lock());
```

Comment (NK-Nikunj) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp]: OK. I just tried compiler explorer to check how compiler evaluates `!try_lock()` and it worked as well. I was thinking that it will lead to the evaluation of exchange on each iteration and thus avoided writing that way.

**Comment by NK-Nikunj** [comment]
@hkaiser @msimberg Not the most extensive of tests, but I wrote a small [benchmark](https://gist.github.com/NK-Nikunj/7bf4641ff77b4db1ee3aaf841aa52bfb) and here are the results:

For the unimproved implementation, I got the following results:
```
Minimum Time: 0.862814
Maximum Time: 0.923062
Avg Time: 0.885118
```

For the improved implementation, I got the following results:
```
Minimum Time: 0.847032
Maximum Time: 0.89873
Avg Time: 0.87345
```

As I suspected, the improvements aren't anything big. 1.3% improvements are small enough to even hide within noise but hey... An improvement is still an improvement!

**Comment by hkaiser** [comment]
@NK-Nikunj the numbers you showed are for sequential execution (i.e. low contention). Would you able to do the same for a parallel loop (i.e. high contention)?

**Comment by NK-Nikunj** [comment]
@hkaiser I think I updated the code after I did parallel execution. The execution times above are for the parallel (contended) cases. The sequential results are as follows:

For the unimproved implementation:
```
Minimum Time: 10.0699
Maximum Time; 10.093
Avg Time: 10.0811
```

For the improved implementation:
```
Minimum Time: 10.0697
Maximum Time; 10.093
Avg Time: 10.079
```

The difference in execution times is barely anything over 100 iterations. I suspected a slight increase in execution times due to the additional load operation alongside exchange that we already do but seems like there isn't any difference in execution times.

**Comment by msimberg** [comment]
> As I suspected, the improvements aren't anything big. 1.3% improvements are small enough to even hide within noise but hey... An improvement is still an improvement!

I feel obliged to post this since I just rewatched it yesterday: https://www.youtube.com/watch?v=koTf7u0v41o. Even with that, I really wouldn't be surprised if that's a real improvement, no matter how small.

File: libs/core/thread_support/include/hpx/thread_support/spinlock.hpp [hunk 4]
```diff
@@ -45,10 +46,11 @@ namespace hpx { namespace util { namespace detail {
         void lock() noexcept
         {
             // Optimistically assume the lock is free on the first try
-            if (!m.exchange(true, std::memory_order_acquire))
+            if (try_lock())
```
Comment (msimberg) [inline libs/core/thread_support/include/hpx/thread_support/spinlock.hpp:49]: Is there now any reason why this spinlock should first do `try_lock` and the other one would first check `is_locked`? What's different?

Comment (NK-Nikunj) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp:49]: There's no difference. The first one does a `try_lock` in hopes of getting the lock. If the `load` operation returns false it continues executing without doing an exchange (that we are trying to reduce). The second one does exactly the same. It looks different because `yield_k` functions are different in both implementations. The first one takes a counter k and implements a backoff strategy. The second one takes a predicate and does what the first one does but only until the predicate returns true. So, the design for both of them is essentially the same. The function signature causes them to look different.

Comment (msimberg) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp:49]: Let me rephrase: Why not structure this exactly the same way as the spinlock in `synchronization`, with `is_locked` etc.?

Additionally, with the change to use `try_lock` in the `if` these two are identical:
```
if (try_lock) return;
unsigned k = 0;
do { yield_k(k++); } while (!try_lock());
```
and:
```
unsigned k = 0;
while (!try_lock) { yield_k(k++); }
```

Comment (NK-Nikunj) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp:49]: Aah, right! I got what you're saying now. Yes, it can be structured this way and I should do it do, apologies.

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

File: libs/core/thread_support/include/hpx/thread_support/spinlock.hpp (hunk 4 above)
Comment (msimberg) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp]: Sorry @NK-Nikunj for beating this topic to death, but:

`synchronization/spinlock.hpp` was changed from
```
            while (!acquire_lock())
            {
                util::yield_while([this] { return is_locked(); },
                    "hpx::lcos::local::spinlock::lock");
            };
```
to
```
            do
            {
                util::yield_while([this] { return is_locked(); },
                    "hpx::lcos::local::spinlock::lock");
            } while (!acquire_lock());
```
because of `<reasons>`.

`thread_support/spinlock.hpp` is now:
```
            unsigned k = 0;
            while (!try_lock())
            {
                yield_k(k++);
            };
```

Now my original question in this thread becomes a bit more obvious: Are these two still equivalent, and if not, why should they not be?

In https://github.com/STEllAR-GROUP/hpx/pull/5445#discussion_r669503863 I merely wanted to point out that if there is going to be an `if (x) return; do { z; } while (!x);` it's better written as `while (!x) { z; };` not necessarily suggest that it should be a `while (!x) { z; };`

Comment (NK-Nikunj) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp]: > Are these two still equivalent, and if not, why should they not be?

Yes. I think the confusion arises from the behavior of yield functions. Let's consider both cases one by one. For `thread_support/spinlock.hpp`, we do:

```
unsigned k = 0;
while (!try_lock())
{
    yield_k(k++);
};
```

This does one atomic load which when computed to true does a yield of 0 following which it does another atomic load. If that computes to true as well then we do yield of 1 and so on. In essence, we are doing atomic loads as long as yield either suspends the thread or the atomic loads computes to false. Finally, if it computes to false, it does an atomic exchange to try and acquire the lock.

Now consider `synchronization/spinlock.hpp` file with `do {} while` implementation. Here, it goes into yield first (the internal yield variable k is still 0) and tries for an atomic load (defined by `is_locked()`). It continues to yield with k = 1, 2, ... until `is_locked` returns false following which it does an atomic exchange to acquire the lock (defined by `!acquire_lock()`). So, it has the same functioning as the above case.

Now consider the case with `while` instead of `do {} while`. Here we do an extra atomic load in the while loop. So, for k = 0, we do one atomic load when we enter the loop followed by another atomic load through `is_locked()`. Now, this wouldn't matter for scenarios where the first thread acquires the lock (as we never get into the while loop) but when the thread has to spin on while, we end up doing an extra load.

Again, the cost of a single atomic load might not be substantial, but I won't call them equivalent implementations then :P

Comment (hkaiser) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp]: We should avoid unneeded atomic operations in any case. They might create non-measurable overheads, but they are still atomics (even if if they use relaxed semantics), i.e. they force the core's execution pipeline to flush/stall.

**Comment by hkaiser** [comment]
retest lsu

**Comment by hkaiser** [comment]
The remaining issues are unrelated.

File: libs/core/thread_support/include/hpx/thread_support/spinlock.hpp (hunk 4 above)
Comment (msimberg) [reply libs/core/thread_support/include/hpx/thread_support/spinlock.hpp]: For posterity, we decided to attempt the unification of the two spinlock implementations separately. This is resolved for now.
